### ISSUE TITLE
Hard override to resolve situations where consensus is lost

### ISSUE THREAD

**Issue by timmaxw** [rethinkdb/rethinkdb#3893]
Sometimes the user may lose more than half of the voters in their Raft cluster or more than half of the replicas for a shard of a table. For example, two hard drives might fail at the same time; or a tornado might wipe out an entire datacenter. Currently this situation is unrecoverable. We should give the user a way to manually override the normal quorum requirements to recover from this situation. If the user decides to use these manual overrides, our normal consistency guarantees are invalidated; we will make a "best effort" to remain consistent, but it might not work.

There are actually two types of hard override:

* We lose half or more of the voters for the Raft cluster for a table, so we cannot execute Raft transactions. In this case, the `table_meta_manager_t` needs to start a new table epoch, so that it can initialize the Raft state to a different set of voters. We need to make an effort to copy the old Raft state into the new Raft cluster, but it's always possible that some changes may be committed on the old Raft cluster and not the new one; we need to not crash in this case.

* We still have a quorum of Raft voters, but we lose half of more of the voting replicas for a specific shard of a table. In this case we can fix it by executing a Raft transaction that overwrites the existing contract with a new one with a different set of voters.

We should also figure out how to present these options to the user. Here are a couple of options:

* Allow the user to write to the `rethinkdb.table_config` system table with a special flag `override=true`.

* Give the user a new command `table.recover()`

I'm inclined to prefer the second option. It gives the implementation more leeway to recover in the safest possible way, instead of overriding the table metadata to the exact state the user specifies. But I haven't thought through the details.

**Comment by timmaxw** [comment]
I'm updating this issue to reflect offline discussion over the past few weeks.

There are three operations we can potentially do to fix a broken shard of the table:

* Demote one or more voters to non-voters, but leave at least one valid voter.

* Demote all of the voters to non-voters and promote one or more non-voters to voter.

* Demote all of the voters to non-voters and promote one or more unrelated servers to voter. (This will cause the loss of all data for the shard.)

After the user has done one of these three operations, they can use the regular reconfiguration mechanism to follow up. For example, they might use the hard override mechanism to demote a voter to non-voter, then use a regular reconfiguration to remove it completely and add a new voter to replace it.

The operations that can be done via hard override are deliberately constrained to be as safe as possible. For example, it wouldn't be safe to demote one or more voters to non-voter and promote one or more unrelated servers to voter in the same step; this might result in the system discarding the data on the existing voters in favor of the lack-of-data on the new voters. This is why we force users to use the regular reconfiguration mechanism to do the second step: it doesn't suffer from the same problem.

However, this causes a problem with the `.update(new_config, override=True)` syntax: if the user is trying to run hard override on a table that's stuck in the middle of a reshard, it's sometimes difficult or impossible to express the new configuration correctly. For example, suppose the user has a table with two shards, one on server A and one on server B. They reconfigure the table so that there is only one shard, on server C. At the moment when both shards are in the middle of the hand-over to add server C as a new replica (so `temp_voters server C goes down. Now the table

**Comment by timmaxw** [comment]
Note for email readers: the above comment has been edited, because I accidentally hit "Submit" in the middle of writing it.

**Comment by timmaxw** [comment]
If we wanted to give the user more control, we could add a `dead_servers=[...]` parameter to `table.reconfigure(emergency_recovery=True)`, to allow the user to specify exactly which servers should be demoted. This would make it easier to explain the behavior to people, perhaps.

Another idea is that we might want to introduce a new command like `table.repair()` instead of overloading `table.reconfigure()`.

**Comment by mlucy** [comment]
Since this is a rare operation, an optarg to `reconfigure` seems better to me than a whole new command.

**Comment by coffeemug** [comment]
:+1: for optarg to `reconfigure`.

**Comment by timmaxw** [comment]
OK, I'm implementing a command of the following form:
```
table.reconfigure(hard_override=True, allow_data_loss=<bool>, dry_run=<bool>)
```
The return value looks just like the normal return value for `reconfigure` except that instead of saying `reconfigured: 1` it might say something different.

What should the optargs be called? It's important that the `hard_override` optarg make it clear how dangerous this operation is. Perhaps `unsafe_repair`, `emergency_repair`, or `unsafe_override`? I think the `allow_data_loss` optarg speaks for itself.

Currently I'm not planning to allow `db.reconfigure(hard_override=True)`. Does that sound reasonable?

**Comment by coffeemug** [comment]
I think we should merge `hard_override` and `allow_data_loss` into one optarg (e.g. `emergency_repair=False|'unsafe'|'catastrophic'`). I think these names are actually quite good, but I'm not particularly attached to them in case someone thinks of better alternatives.

**Comment by danielmewes** [comment]
`emergency_repair=False|'unsafe'|'catastrophic'`
That sounds generally fine, though I would suggest two modifications:
* `'with_data_loss'` instead of `'catastrophic'` as the third value. That is more explicit as to what that setting actually does.
* call the parameter `override` instead of `emergency_repair`, for the simple reason that I think it reads better in case the value is `'unsafe'`. So it would be `override=False|'unsafe'|'with_data_loss'`

There might be a better term for `'unsafe'` as well, but I think `'unsafe'` is probably fine (`'with_rollback'`/`'allow_rollback'` maybe because it might roll back acked writes?).
It's also not relevant if we're only going to support `'with_data_loss'` for now.

Only allowing `'with_data_loss'` overrides sounds like a reasonable simplification for this. I'll think about it a little bit more, but I tend to think that it's fine for the first version.

**Comment by coffeemug** [comment]
I agree in general, but I think `override` isn't a good name because it implies that the user specifies a config that will override what's in the cluster. But we don't let users specify anything, so they're not really "overriding", they're "repairing", which is why I think `repair` or `emergency_repair` is a slightly better name.

**Comment by timmaxw** [comment]
I'd like to support both modes even in the first implementation. There's a huge difference in how dangerous they are.

I agree with @coffeemug that `override` is not a good name.

I think it's important to have the word "unsafe" or a similar word somewhere in every invocation of the command. A user who has not read the documentation should be able to look at the command and immediately see that it's unsafe. I'm not sure that `table.reconfigure(emergency_repair="allow_rollback")` makes that clear enough.

**Comment by coffeemug** [comment]
What about `repair=False | 'unsafe_rollback' | 'critical_rollback'`?

**Comment by timmaxw** [comment]
`critical_rollback` doesn't mean anything to me. What about `unsafe_rollback_or_erase`?

**Comment by coffeemug** [comment]
:+1:

**Comment by coffeemug** [comment]
BTW, these names are a bit long and ugly, but I think in this case it's ok (because the user is asking to do an ugly thing).

**Comment by danielmewes** [comment]
@timmaxw Oh I mis-interpreted 
> Currently I'm not planning to allow db.reconfigure(hard_override=True). Does that sound reasonable?

as meaning that you couldn't do an override without `allow_data_loss=True`, but I guess you meant that `allow_data_loss` just has to be specified either `False` or `True` explicitly?
Assuming I got it right now, please disregard my comment about not supporting the equivalent of `allow_data_loss=False`. That being said I think `allow_data_loss=False` is misleading because acknowledged writes could still get lost as far as I can tell (or am I wrong about that)?

Ok, how about `emergency_repair=False | 'unsafe_rollback' | 'unsafe_erase'`?

I think `repair=False` is weird since you can often use `reconfigure` to "repair" something without needing hard override. With the new names for the possible options I think `emergency_repair` works better.

**Comment by VeXocide** [comment]
> What about `repair=False | 'unsafe_rollback' | 'critical_rollback'`?

I still dislike the mixing of types, while we're doing it in one or two places already in my opinion it's something we should avoid.

**Comment by timmaxw** [comment]
When I said "Currently I'm not planning to allow `db.reconfigure(hard_override=True)`" I meant that the user would have to run the command on each table one at a time, rather than on the entire database at once. The important part is the `db.` versus `table.`.

I dislike `unsafe_erase` because it will sometimes rollback. `unsafe_rollback_or_erase` is a much more accurate description of what it does.

I agree that `allow_data_loss` is a bad name because it incorrectly implies that `allow_data_loss=False` will never lose data.

**Comment by coffeemug** [comment]
Well, `False` wouldn't really be a type (we don't have to allow the boolean -- it could be two strings or nothing).

**Comment by timmaxw** [comment]
:+1: for `emergency_repair='unsafe_rollback'` or `emergency_repair='unsafe_rollback_or_erase'` or omitting the optarg entirely. I agree we should reject `emergency_repair=False`.

**Comment by danielmewes** [comment]
> `emergency_repair='unsafe_rollback'` or `emergency_repair='unsafe_rollback_or_erase'` or omitting the optarg entirely. I agree we should reject `emergency_repair=False`.

Sounds fine. I have no strong opinion on whether or not to allow `False` except that it should be
a) consistent with what we do elsewhere and
b) that there should be *some* value that you can pass in to mean the default (could be `nil` or `False`)

I think b) is a good idea because it makes scripting `reconfigure` queries easier. Say you write an ORM and encapsulate the call to `reconfigure` into a method on the schema object. If the only way to specify the default behavior is by omitting the `emergency_repair` optarg completely, you might need to distinguish two cases somewhat like this:
```
function reconf(config, repair = UNDEFINED) {
    if (repair == UNDEFINED)
        table.reconfigure(config).run(conn);
    else
        table.reconfigure(config.merge({emergency_repair: repair})).run(conn);
}
```
Otherwise you can just do
```
function reconf(config, repair = false) {
        table.reconfigure(config.merge({emergency_repair: repair})).run(conn);
}
```

So I think we should allow either `null` or `false` as a value to make it equivalent to the default.

**Comment by mlucy** [comment]
We should have a way of passing the optarg's default value for consistency with the rest of the query language.  Every other optarg has a default value you can pass.  I'm not too concerned about mixing types here, ReQL is a dynamic language.

(Also, it seems people are happy with the current solution, but what about `allow_unsafe=False` `allow_unsafe='rollback'` and `allow_unsafe='data_loss'`?  That reads pretty clearly to me.)

**Comment by VeXocide** [comment]
In that case I'd prefer `foo=None, foo='bar', foo='baz'`, like an optional.

**Comment by timmaxw** [comment]
I'd be fine with `emergency_repair=None|'unsafe_rollback'|'unsafe_rollback_or_erase'`.

The problem with `allow_unsafe` is that it doesn't say what's actually happening. `reconfigure(emergency_repair="unsafe_rollback")` is fundamentally different from `reconfigure(shards=X, replicas=Y)`. That's why I like having the words `emergency_repair` in there; it's effectively the name of a subcommand you're invoking. But `allow_unsafe="rollback"` makes it sound like you're modifying the normal command, which you're not.

**Comment by VeXocide** [comment]
As for the names, I like `emergency_repair='unsafe_rollback'` and `emergency_repair='unsafe_rollback_or_erase'` but I have no strong opinion on the name.

**Comment by danielmewes** [comment]
Well there is no `None` value in ReQL, so that would be the equivalent of `null` in whatever language is used.

@mlucy Do you have a good overview of what we use elsewhere?

Apart from this last question of whether to use `null` or `false` for the default, or even allowing both, it seems that we agree on
`emergency_repair=X|'unsafe_rollback'|'unsafe_rollback_or_erase'`

**Comment by VeXocide** [comment]
Sorry, I obviously meant `null`.

**Comment by danielmewes** [comment]
Alright, let's do
`emergency_repair=null|'unsafe_rollback'|'unsafe_rollback_or_erase'`
then.

**Comment by timmaxw** [comment]
In CR 2939 with the API we most recently settled on.

We should consider renaming `rollback` to `demote` perhaps.

**Comment by danielmewes** [comment]
"demote" doesn't really say what can happen to the user's data does it?

**Comment by timmaxw** [comment]
Fixed in 81ca7786ef849b0fe5d082157ff1023a048ae4cf.
